### ISSUE TITLE
DownloadManager.submitAllStoredDownloads swallowing events

### ISSUE THREAD

**Issue by MechaRyRy** [novoda/download-manager#454]
## Problem 
It seems like that any call to the `submitAllStoredDownloads` is preventing further event calls, pause etc, from going through. I'm assuming that the reference we store in the internal map is being nuked on subsequent calls which leads to unexpected behaviour. 

#### Steps to reproduce
- Open the demo application and start downloading
- Spam the "Download only over wifi button"
- Try to pause and resume downloads / delete them

**Expected:** actions work as expected
**Actual:** It deletes downloads and then immediately re-queues them (this is probably due to adding the same reference to the executor multiple times)
**Actual:** It doesn't pause / resume the download.

## Solution
- Add additional logs around this `submitAllStoredDownloads` and the references accessed when performing events. 
- Find out where this problem is originating and patch the issue.
- I think we should drop `submitAllStoredDownloads` from the public API, calls to this method should only be handled internally. 

## Impact
This is a serious issue, we could be downloading multiple times, not removing downloads etc, this is a huge impact to our client applications.

**Comment by MechaRyRy** [comment]
This doesn't seem to be tied to the job that reschedules because I can reproduce if I always remain in mobile network (making it never execute).

**Comment by Lingviston** [comment]
Today I encountered the same bug, while playing with queue management.
My steps to reproduce (I used modified demo with 5 files instead of 2):
1. Start downloading of several files;
2. Verify that they're queued;
3. Verify that first file started downloading;
4. Fail downloading of the first file;
5. Verify that second file started downloading;
6. Pause or delete second file

**Expected**: Second file gets paused or deleted;
**Actual**: Second file continues downloading.

This happens because Novoda supports auto-restoration after connection lost. However ANY connection error is treated by Novoda as connection lost. The mechanism is the following: whenever connection error happens Novoda fetches the downloads from persistence storage and submits them to download. Those fetches downloads are represented by Java instances which are different from the initial ones. In the described scenario when user tries to pause or delete the second download (which is ongoing) after the failed first one it tries to manipulate with the "old" instance, while internally Novoda has already rewritten link to it with the new one.

I managed to fix it by removing
`downloadBatchMap.put(downloadBatch.getId(), downloadBatch);`
line from LiteDownloadManager#loadBatchesCallback and modifying the LiteDownloadManagerDownloader#download in the following way:
```
void download(DownloadBatch downloadBatch, Map<DownloadBatchId, DownloadBatch> downloadBatchMap) {
        DownloadBatchId downloadBatchId = downloadBatch.getId();
        if (!downloadBatchMap.containsKey(downloadBatchId)) {
            downloadBatchMap.put(downloadBatchId, downloadBatch);
        }

        Logger.d("Downloader download called for " + downloadBatch.toString() + " with id " + downloadBatch.getId().rawId());
        executor.submit(new Runnable() {
            @Override
            public void run() {
                Wait.<Void>waitFor(downloadService, waitForDownloadService)
                        .thenPerform(executeDownload(downloadBatch, downloadBatchMap));
            }
        });
    }
```

The idea is that we should not put new batches into map if batches with the same id are already there. I'm not sure how solid is mine solution as I performed only smoke test, so no pull request.

**Comment by MechaRyRy** [comment]
Hi @Lingviston, I figured it would be something with the maps. I don't see how:

```
void download(DownloadBatch downloadBatch, Map<DownloadBatchId, DownloadBatch> downloadBatchMap) {
        DownloadBatchId downloadBatchId = downloadBatch.getId();
        if (!downloadBatchMap.containsKey(downloadBatchId)) {
            downloadBatchMap.put(downloadBatchId, downloadBatch);
        }

        Logger.d("Downloader download called for " + downloadBatch.toString() + " with id " + downloadBatch.getId().rawId());
        executor.submit(new Runnable() {
            @Override
            public void run() {
                Wait.<Void>waitFor(downloadService, waitForDownloadService)
                        .thenPerform(executeDownload(downloadBatch, downloadBatchMap));
            }
        });
    }
```
this has changed though, it looks the same to me. Are you saying just to update the `LiteDownloadManager#loadBatchesCallback` and it solves it?

You would still be queuing the same instance of the `DownloadBatch` against the executor. Now however, the map would only ever represent the original `DownloadBatch` which is not an issue assuming that it is mutated correctly.

**Comment by Mecharyry** [comment]
The problem is `pause` / `resume` events are still not going through because

**Comment by Lingviston** [comment]
@Mecharyry sorry, I pasted wrong snippet of code. I'll provide correct code snippet for the second change on Monday. Or maybe even do a PR.

I don't think that the first change is enough. In the use-case that I described - it might be (and I've proven it via smoke test) but I could imagine some scenaios where both bunches of changes are required.

**Comment by Lingviston** [comment]
Here's a corrected code snippet:
```
void download(DownloadBatch downloadBatch, Map<DownloadBatchId, DownloadBatch> downloadBatchMap) {
        DownloadBatchId downloadBatchId = downloadBatch.getId();
        DownloadBatch submittedBatch = downloadBatchMap.get(downloadBatchId);

        final DownloadBatch batchToDownload;
        if (submittedBatch == null) {
            downloadBatchMap.put(downloadBatchId, downloadBatch);
            batchToDownload = downloadBatch;
        } else {
            batchToDownload = submittedBatch;
        }

        executor.submit(() -> {
            Wait.<Void>waitFor(downloadService, waitForDownloadService)
                    .thenPerform(executeDownload(batchToDownload, downloadBatchMap));
        });
    }
```

**Comment by Lingviston** [comment]
@Mecharyry thank you for addressing it so quickly! I saw you created a PR, but later on it has been just closed. Does it make sense if I create it by myself? I wanted to contribute:)

**Comment by MechaRyRy** [comment]
@Lingviston you can always pull the branch and add some tests if you wish? 😄 I won't have time until later today. Your choice!

**Comment by MechaRyRy** [comment]
@Lingviston if you are interested you can test the fix we just did using the `snapshots` as detailed in the readme -> https://github.com/novoda/download-manager we'll try and get this released as soon as possible after taking a look at some other bugs.

**Comment by Lingviston** [comment]
@Mecharyry thanks. Fix is based on the changes which I already checked. Looking forward to get a new release soon.

**Comment by MechaRyRy** [comment]
Released in `2.2.0`
